eventhubs/client/azure-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubConsumer.java

```diff
@@ -0,0 +1,121 @@
+    private final AtomicBoolean isDisposed = new AtomicBoolean();
+    private final ClientLogger logger = new ClientLogger(EventHubConsumer.class);
+    private final EmitterProcessor<EventData> emitterProcessor;
+    private final Flux<EventData> messageFlux;
+
+    private volatile PartitionProperties partitionInformation = null;
+    private volatile AmqpReceiveLink receiveLink;
+
+    EventHubConsumer(Mono<AmqpReceiveLink> receiveLinkMono, EventHubConsumerOptions options, Duration operationTimeout) {
+        this.receiveLinkMono = receiveLinkMono;
+        this.emitterProcessor = EmitterProcessor.create(options.prefetchCount(), false);
+        this.operationTimeout = operationTimeout;
+
+        // Caching the created link so we don't invoke another link creation.
+        this.messageFlux = receiveLinkMono.flatMapMany(link -> {
+            if (RECEIVE_LINK_FIELD_UPDATER.compareAndSet(this, null, link)) {
+                logger.asInfo().log("Created AMQP receive link. Initialising prefetch credit: {}", options.prefetchCount());
+                link.addCredits(options.prefetchCount());
+
+                link.setEmptyCreditListener(() -> {
+                    if (emitterProcessor.hasDownstreams()) {
+                        return creditsToRequest.get();
+                    } else {
+                        logger.asVerbose().log("Emitter has no downstream subscribers. Not adding credits.");
+                        return 0;
+                    }
+                });
+            }
+
+            return link.receive().map(EventData::new);
+        }).subscribeWith(emitterProcessor);
+
+        emitterProcessor.doOnSubscribe(subscription -> {
+            if (receiveLink.getCredits() == 0) {
+                logger.asInfo().log("Subscription received and there are no remaining credits on the link. Adding more.");
+                receiveLink.addCredits(creditsToRequest.get());
+            }
+        }).doOnRequest(request -> {
+            logger.asInfo().log("Back pressure requested. Old value: {}. New value: {}", creditsToRequest.get(), request);
+            creditsToRequest.set((int) request);
+        });
+    }
+
+    /**
+     * Disposes of the consumer by closing the underlying connection to the service.
+     *
+     * @throws IOException if the underlying transport and its resources could not be disposed.
+     */
+    @Override
+    public void close() throws IOException {
+        if (!isDisposed.getAndSet(true)) {
+            final AmqpReceiveLink receiveLink = RECEIVE_LINK_FIELD_UPDATER.getAndSet(this, null);
+            if (receiveLink != null) {
+                receiveLink.close();
+            }
+
+            emitterProcessor.dispose();
+        }
+    }
+
+    /**
+     * Gets the most recent information about a partition by the current receiver.
+     *
+     * @return If {@link EventHubConsumerOptions}
+     */
+    public PartitionProperties partitionInformation() {
```

Comment:
I would argue that this should be 'getPartitionInformation'

Reply:
Thanks for pointing that out, it is a post-preview 1 feature. Removed.